wasmplugin/plugin.go

```diff
@@ -34,43 +34,62 @@ func (*vmContext) NewPluginContext(contextID uint32) types.PluginContext {
 	return &corazaPlugin{}
 }
 
-type corazaPlugin struct {
-	// Embed the default plugin context here,
-	// so that we don't need to reimplement all the methods.
-	types.DefaultPluginContext
-	wafSets wafSets
+type wafMap struct {
+	kv         map[string]coraza.WAF
+	defaultKey string
+}
 
-	wafDefaultEnabled   bool
-	wafDefaultDirective string
+func newWAFMap(capacity int) wafMap {
+	return wafMap{
+		kv: make(map[string]coraza.WAF, capacity),
+	}
+}
 
-	perAuthorityWafSets perAuthorityWafSets
-	metricLabels        map[string]string
+func (m *wafMap) put(key string, waf coraza.WAF) error {
+	if len(key) == 0 {
+		return errors.New("empty key")
+	}
 
-	metrics *wafMetrics
+	m.kv[key] = waf
+	return nil
 }
 
-type wafSets []wafSet
+func (m *wafMap) setDefaultKey(key string) error {
+	if len(key) == 0 {
+		return errors.New("empty default key")
+	}
 
-func (wfs wafSets) newWAFSetsHttp(contextID uint32) wafSetsHttp {
-	var wafSetsHttp wafSetsHttp
+	if _, ok := m.kv[key]; ok {
+		m.defaultKey = key
+		return nil
+	}
 
-	for _, wafSet := range wfs {
-		var wafSetHttp wafSetHttp
-		wafSetHttp.tx = wafSet.waf.NewTransaction()
-		wafSetHttp.logger = wafSet.waf.NewTransaction().
-			DebugLogger().
-			With(debuglog.Uint("context_id", uint(contextID)))
-		wafSetHttp.name = wafSet.name
+	return fmt.Errorf("unknown key %q", key)
+}
 
-		wafSetsHttp = append(wafSetsHttp, wafSetHttp)
+func (m *wafMap) getWAF(key string) (coraza.WAF, error) {
+	if w, ok := m.kv[key]; ok {
+		return w, nil
 	}
 
-	return wafSetsHttp
+	return m.getDefaultWAF()
```

Comment:
in this cases, we will always enable WAF?

Reply:
what about if I only want to enable in foo.example.com? is that still possible?

Reply:
this is the previous code behaviour

| **default_directive** | **per_authority_directives** | Expectation |
|---|---|---|
| global_directive | "foo.example.com": "foo_directive" | 1. foo.example.com will implement foo_directive<br>2. any other authority will implement global_directive  |
| global_directive | N/A | 1. all traffic will implement global_directive  |
| N/A | "foo.example.com": "foo_directive" | 1. foo.example.com will implement foo_directive<br>2. ignore any other traffic |
| N/A | N/A | 1. ignore all traffic |

Reply:
cc @jcchavezs

Reply:
ignore it, seems like it's cover in getDefaultWAF()